contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,92 @@
+# Local PV Distributed Static Provisioner
+Authors: @verult, @msau42
+        
+## Design Goals
+1. Limit all PV object interactions to a single master pod in order to minimize node access to the Kubernetes system.
+1. Assume disk deletion is asynchronous, as required by block volumes.
+
+## Related Works
+* Original provisioner design, [here](https://github.com/kubernetes-incubator/external-storage/tree/master/local-volume/provisioner#design).
+* Disk cleaner (deleter) implementation for block volumes, [here](https://github.com/kubernetes-incubator/external-storage/pull/312).
+
+## Design 1: Master-Heavy
+
+### High Level Overview
+The static provisioner will be split into a single master (ideally placed on the master node) and many minions (one per node). The master contains the bulk of provisioner logic, including syncing PV objects with the API server and triggering discovery and deletion. Minions are left to perform the physical discovery and deletion of the local volumes on its node. gRPC is used to communicate between master and minions, with the master as the client and each minion as a separate server.
+
+The master keeps track of `MinionInfo`, a collection containing one entry for each minion keeping track of relevant information. The master also runs a group of goroutines, one for each minion, which controls and interacts with the minions.
+
+### Minion Provisioner
+The minion provisioner exposes the following RPCs:
+
+| RPC                           | Description |
+| --- | --- |
+| `GetMinionName()`             | Returns the minion provisioner's name. |
+| `GetMinionStatus()`           | Returns the minion status, as described in the next section. |
+| `Delete(PersistentVolume)`    | Starts the asynchronous delete of the given PV. |
+
+In addition, the Discoverer and Deleter from the previous design are kept here, but they only perform filesystem operations. All other logic is moved to the master.
+
+### MinionStatus
+The following fields are retrieved as part of the `GetMinionStatus()` call:
```

Comment:
How often is `GetMinionStatus` polled by master? Would it make sense to add `Watch()` call that would return `stream Events`? It would either poke the master that it should call `GetMinionStatus` or the event could even contain what has changed so the master does not need to call `GetMinionStatus` over and over again.

Reply:
Revisiting gRPC designs since a hybrid approach (gRPC + CRD) looks more promising.

@jsafrane If the number of nodes is large, would there be a scalability problem with a large number of streams?

Reply:
You may hit the default process limit of 1024 file descriptors.

Reply:
In order to scale with # nodes, we will not be able to keep long running connections open.  We may need to open a new connection with every request, which is not great if requests are frequent.

Reply:
Do we need long running connections here? I think it makes the most sense for the minion to initiate the chatter by poking the server (as jsafrane suggested). It will make sure interactions are only driven by actual events and not regular polling. So the process could look like this:

1) Minion has some update to report, connects & requests a callback from master and disconnects.
2) This request gets queued up in the master for processing. 
3) Master calls back to get the information and updates CRD and disconnects.

The number of go routines pulling things off the work queue can be limited to any file descriptor limits. In general I believe it should be easy for a process to set the ulimit within a container to whatever fd limit it wants, so this limit could be high. But in general, I don't think we need long running connections for infrequent state changes in volume's cleaning and recycling status.

Reply:
That's fine, we just can't use rpc streams then because that requires the connection to stay open.  If minion initiates the connection, then that brings up the issue of how to verify minion identity in case minion tries to spoof another node.

Reply:
Is it correct to say even if a faulty minion spoofs an IP, if master sends a request to that IP, it'll still be routed to the legitimate minion?

@dhirajh In addition, minions should keep poking master until it receives a status request, to make sure master gets a status update. It can also be less aggressive and only retry if there was a network failure. And as a safety measure, master should still periodically poll for status, but very infrequently.

Reply:
@verult @msau42 No. The above measures only safeguards against a random client with arbitrary network identity claiming to be the POD. If the network communication layer has been compromised to the extent that packets meant for the POD are re-routed to a fake POD then these measures won't work. To guard against such an attack, one would need minion authentication (just like PKI in Kubelets). Using any authentication mechanism which certifies each minion uniquely will probably require a non-DaemonSet approach as each POD will need unique secret. 

How about a different approach to this problem? 
### Job Based Approach
1. We still have a master and minions in each node.
2. The minions have one and only one role to play - They detect new disks and they poke the master.
3. The master acts (and trusts) purely through "Jobs" (https://kubernetes.io/docs/concepts/workloads/controllers/jobs-run-to-completion/).

#### Disk Discovery:
When it gets pinged by a minion about new disk, the master launches a Job which runs on the node (by setting nodeSelector to that hostname). That job could have a secrets volume mounted to effectively authenticate itself with the master as it reports back the newly discovered disk via gRPC or REST. The lifespan of that secrets volume can even be limited to that job run if we want to be very secure.

#### Disk Cleaning/Deleting:
When the master concludes (based on PV state) that disk needs cleaning, it launches a cleaning Job on the node. This Job doesn't even need to report back to the master on state changes.

 - If the Job is still running => Cleaning is still in progress
- If the Job fails => Cleaning failed
- If the Job completes => Cleaning completed.

Note that the minion is not involved in the cleaning process. The master uses the status of the Job to determine if the PV recycling is completed or not. All status is maintained in Kubernetes resources (PVs and Jobs), which is highly desirable. Jobs could be named in a manner that makes it easy to identify which node and disk they are targeting (to improve administrative insight).

Reply:
The Job might not even need credentials. It could print the discovery result to STDOUT, and master can then grab its logs.

Reply:
The Jobs approach looks promising, thanks so much for the idea!  If possible, can we try to eliminate grpc completely from the design?  It would make the design and implementation much simpler if we can stick to one consistent communication mechanism.

At the simplest, you could just use Jobs if the discovery was polling-based.  If we want to support event-based discovery, could a crd work?  If the only field is to trigger discovery, then the worst that could happen we run discovery jobs more frequently than necessary.

Reply:
If the discovery is polling based, then you can eliminate the DaemonSet entirely.

Reply:
As @verult suggested the master could read STDOUT for results (and avoid grpc). STDOUT tends to be brittle because any extraneous message can break parsing. However, given that these are simple jobs that will do one very specific thing and exit, the chances of this happening are very low and they will only happen due to genuine errors in the primary task being carried out.

Regarding polling, I shied away from polling using jobs as the cost at higher frequency was unclear to me. However, if its ok to poll at a low frequency (once every few hours, or day, configurable?) then its low impact. An admin could force polling (because they just made a change to the host) by checking a field - perhaps a label?. We won't even need a CRD for that, the label could be set on the node. The master anyway needs to "watch" all nodes to discover new nodes as they come online.  The protocol could be as follows:
The master sets the time it last polled a node in a node label.
- New nodes won't have that label, so they will get immediately polled. 
- The master uses the label value to decide when next to poll (has it been 2 hrs yet?).
- An Admin can force the polling on a node by deleting the label or setting it to zero.

Reply:
I meant, even if you can read the discovery output from stdout, how does the master know when to launch the job in an event-based model?  I don't think we should modify the node object labels or annotations.  Only kubelet should.

Could stdout potentially get truncated if the output is too long?  I'm not familiar with how the logging mechanism works, but mixing logs with data may not be great.  Can we pass data with a secret?

Reply:
Using job logs is another nice potential avenue for simplifying the overall design. 

If the logs are used, we should leverage some type of structured logging (e.g. https://github.com/sirupsen/logrus and json). Instead of simple less structured log lines (e.g. traditional glog).

Reply:
For achieving event based discovery, one of these two seem possible:

1. Discovery is run as a long running job or regular Pod explicitly by the master.  This Job/Pod can easily be supplied with unique serviceAccount information to update CRDs or use stdout or grpc, because the manifest for them is crafted by the master.  The downside is the need to manually start this Job/pod on every node and be aware of new nodes as they come online, deal with node failure, etc. This seems complicated to me as the master gets into the business of being a Pod controller.

2. Use the DaemonSet approach of deploying Pods. A DaemonSet takes a lot of complexity out  of the equation for the master regarding discovery of new nodes and other issues mentioned in option 1. To your point, yes, we could use CRD for communication here. Because its a DaemonSet, it would use the same serviceAccount on all Pods to update some "notification" CRD (instead of stdout or grpc/REST), but this is just for notification and as you pointed out the ill effect of a security compromise is a more frequent discovery request. Once the master sees the notification, it can launch a job, which could use a unique service account specified in its manifest to update the CRD (instead of stdout or grpc/REST with unique tokens).  The main question I have is how complicated it is to maintain unique service accounts for each node? Would that involve so much complexity that one of the other mechanisms seem more preferable?

Reply:
Agree we don't want to over complicate this.  Adding/removing local disks from a node should be an uncommon occurrence.  One more thing to keep in mind: if it's uncommon enough, is it worth it to have a long running pod always running on every node?  That's more system overhead we are adding in order to handle a rare event.

Reply:
IMO using CRDs only as notifications is an abuse of its original intent. It's meant as a mechanism for a declarative API, not just a way to share state. Between gRPC/REST and CRD for notifications I would prefer gRPC.

Reply:
@verult gRPC would be fine. One thing you might want to iron out is how a notification will be either persisted by the master or retried by the minion (if the master loses it). Keep in mind that the master will need to be capable of being rescheduled on another node if the node it is running on has failed, so it may not want to use local storage for persistence.

Reply:
@jsafrane @childsb can you guys review this Jobs approach considering the issues you guys encountered with recycling?